### ISSUE TITLE
Alternative version of callback-taking functions

### ISSUE THREAD

**Issue by danielkeller** [emilk/egui#1004]
I'm working out how to wrap egui's API for my scripting language. Something that would make this a lot easier is if there was an alternative API to the functions that take callbacks (like `SidePanel::show`, `Window::show`, and `Ui::horizontal`) which could be written in straight-line code. For example, an object that exposes a `&mut UI` to draw widgets, and has a `finish` method which drops it.

It looks like I can emulate this with some pairs of `Ui::new` and `Ui::allocate_rect`, but I'll have to rewrite all the useful stuff those functions do.

**Comment by coderedart** [comment]
I might be wrong, but i think the point of closures is that they will run only if necessary (ui is shown).

With a straight sequence style of begin and end functions, we will have to throw around if shown else not shown kind of code. Or is there a more idiomatic code style?

**Comment by emilk** [comment]
egui passes closures to functions to call, like this:

``` rust
ui.collapsing_header("Foo", |ui| {
    // …
});
```

These aren't really "callbacks" in the classical sense since they are called immediately (or never).

The reason for them is that the parent `Ui` must run some setup-code and then some tear-down code after the child has been layed out. In the code above there are two different `Ui` both with the name `ui` (parent/child), but since we only ever care about the innermost, this is a good use of shadowing.

---

One alternative could be this (used e.g. by Dear ImGui and Nuklear):

``` rust
if ui.collapsing_header_begin("Foo") {
    // …
}
ui.collapsing_header_end();
``` 

This has two downsides:

A) it is way to easy to put the `end` call in the wrong place (or forget it completely), leading to a lot of subtle bugs that only show up a t runtime.
B) it requires the library (egui) to keep a manual stack of what is going on

We could of course consider adding support for a lower lever API like this, but due to B) above it would require a huge rewrite and redesign of egui.

---

Another alternative us using linear types (types you must manually close):

``` rust
if let Some(child) = parent.collapsing_header("Foo") {
    // …
    child.end(parent); // must call, or it won't compile
}
``` 

Linear types aren't supported by Rust (though you can approximate them with hacks that yields linker errors if you fail to close them), but it also creates a lot of `child`/`parent` bindings, which can be confusing.

---

If Rust had something like Pythons `with` statements we could do:

``` rust
with ui = ui.collapsing_header("Foo") {
   // …
}
```

and have the compiler insert the `end` call automatically. But Rust doesn't support anything like this.

**Comment by danielkeller** [comment]
Thanks for the explanation! What about something like RAII objects? This seems to be the pattern the standard library uses to run tear-down code, for example unlocking mutexes.

**Comment by parasyte** [comment]
It might be ok to impl `Drop` on structs returned by parents. The only catch is that Rust doesn't guarantee destructors are always called. This is why e.g. [`crossbeam::scope`](https://docs.rs/crossbeam/latest/crossbeam/fn.scope.html) takes a callable. For any cases where _not calling_ the destructor would result in unsoundness, this is currently the best pattern offered by the language.

**Comment by emilk** [comment]
A RAII approach would also require a `Ui` to keep a reference to its parent `Ui`, leading to a lot of hassle with lifetimes.

I would be more interested in investigating a more low-level C-like API (with begin/end) and then build the current `Ui` interface on top of that as a convenience. But that's a lot of work. It also is not obvious how one would handle third-party containers.

**Comment by emilk** [comment]
Another downside of the current closure approach is code bloat and long compile times (due to excessive monomorphization)

**Comment by lunixbochs** [comment]
+1, I want to bind egui to python but the callback approach is much worse than something I can wrap in a context manager

I'm imagining this to prototype it:
```
@ui.collapsing_header
def cb(ui):
    ...
```

but I'd really like to be able to do something like this (which I confirmed properly nests scope in Python without persistently shadowing the outer ui variable)
```
with ui.collapsing_header() as ui:
    ...
```

My vote is to try this with Drop:
```
if let Some(ui) = ui.collapsing_header() {
    ...
}
```

I looked into `Drop` a bit and it seems like it should be stable enough for RAII in most cases: https://aloso.github.io/2021/03/18/raii-guards.html#when-drop-isnt-called

I think `Box::leak` doesn't apply unless you're returning a boxed ui object. Same goes for reference counting cycles.

**Comment by schungx** [comment]
> I'm working out how to wrap egui's API for my scripting language. Something that would make this a lot easier is if there was an alternative API to the functions that take callbacks (like `SidePanel::show`, `Window::show`, and `Ui::horizontal`) which could be written in straight-line code. For example, an object that exposes a `&mut UI` to draw widgets, and has a `finish` method which drops it.

Yes, a fluent API is great for programmers, but not friendly to automation, which is what a scripting system ultimately is.

There should be a low-level, imperative, step-by-step API for machine-controlled environments.

**Comment by lunixbochs** [comment]
This is incomplete, but I'm working on a concept here:

1. Playing around with a Context Manager API: https://gist.github.com/rust-play/8667abe111edeb90bca9a44687ecdca4
2. Working on structurally wiring some of the ui api up to it: https://gist.github.com/rust-play/c618172551bebd4f93d929f870d4baa4

Basically, a method like `ui.horizontal()` would return an object conforming to the `UiContextManager` trait, which will need to run before and after the user's inner UI code. To make a context manager work, we really just need some state and an exit() method (or a Drop impl). Nested context managers will work fine, you just chain the exit() methods.

Children don't need a reference to their parent, the context manager can be a short lived object that has a reference to the parent, the child, the mutable Result object, and its internal state necessary for exit().

The `ui.*` api can be moved to a trait as demonstrated, so you can call ui methods like `ui.horizontal()` on the returned context manager.

**Comment by lunixbochs** [comment]
This approach seems somewhat clean for defining using the context manager at call sites:
https://gist.github.com/rust-play/5dbf90f3c22fcec9ce0de00cb87f4273

If necessary the Context object could be either dynamic or templated, allowing each specific call site to stash arbitrary data.

**Comment by coderedart** [comment]
If there is going to be a partial low level rewrite of egui api,  then we might as well check if the new api will be suitable to write bindings for at the very least in wasm/deno_core/pyo3 or rustpython/mlua.  If the new api is fine for one of these and not for others, it would be really sad. Ideally egui will have a simpler C based abi / ffi compatible as that will be the v⅘

**Comment by lunixbochs** [comment]
> we might as well check if the new api will be suitable to write bindings

Yes, I'm specifically doing this for PyO3. (Technically you could probably already bind the existing API, but it would require a really awkward rust -> language -> rust -> language stack to satisfy the egui callback API.)

---

I've prototyped a new context manager style here: https://gist.github.com/0b550112406de053660238627d3189d9

It allows you to write widgets like this, with both the setup and teardown code in the same function:
```rust
fn horizontal(&'a mut self, render: bool) -> Option<ContextManager> {
    if render {
        let child = Ui{id: self.id + 1, parent: self.id, data: 0};
        let local = 1;
        Some(ContextManager::new(move |parent, child| {
            println!("cm exit parent={:?} child={:?} local={}", parent, child, local);
            child.data += 1;
            parent.data += 1;
        }, self, child))
    } else {
        None
    }
}
```

It can be polyfilled to the old callback API style like this:
```rust
fn horizontal_cb<R>(&mut self, f: impl FnOnce(&mut Self) -> R, render: bool) -> Option<R> {
    if let Some(mut cm) = self.horizontal(render) {
        Some(f(&mut cm.child))
    } else {
        None
    }
}
```

@emilk thoughts?

**Comment by lunixbochs** [comment]
I started a port here: https://github.com/talonvoice/egui/tree/no-callback

I ported everything in ui.rs to my context manager approach, with a `_ctx` method suffix. I modified all of the `add_contents` methods to use the equivalent `_ctx` method internally. At a glance, the egui demo app appears to completely work.

I have two notes:
- I'm managing the lifetime of Response objects using an Option type. The callback compatibility methods end up calling `unwrap()` on the response (at a point in time where I know the Response has been set), because InnerResponse isn't optional.
- My ContextManager assumes only one child UI object, and that assumption broke for columns(), so it's not properly wrapped for now.

**Comment by Starwort** [comment]
> Another alternative us using linear types (types you must manually close):
> 
> ```rust
> if let Some(child) = parent.collapsing_header("Foo") {
>     // …
>     child.end(parent); // must call, or it won't compile
> }
> ```
> 
> Linear types aren't supported by Rust (though you can approximate them with hacks that yields linker errors if you fail to close them), but it also creates a lot of `child`/`parent` bindings, which can be confusing.

If you implement `Drop` for `child` then this could work: [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0489ed57a8963ba9ac6068bc40aa07b0)

**Comment by kablouser** [comment]
3 years on... any developments on this?

**Comment by optozorax** [comment]
I also use egui in such way that I have layout of vertical(horizontal + horizontal + horizontal), and it so happens that I need to open one horizontal in one function and end it inside another function. 

It's because I have different objects with different responsibilities, that generate one UI. ~~So, I guess for my use-case, linear types wouldn't fit, because I don't know if this would be possible for the compiler to verify that the linear type that I have passed would always be closed, because I close this inside object that implements some trait.~~

This is why I stuck with `horizontal_wrapped` with `end_row`, it simulates my case very well. But have some bugs that I'm trying to resolve in https://github.com/emilk/egui/pull/8000. ~~So, for me I would only need `begin_horizontal`/`end_horizontal` + `begin_vertical`/`end_vertical`. I don't know about other things, they maybe too general.~~ Maybe it's possible to mark them as `unsafe` and panic if order of opening and closing is incorrect? I would be OK with such design.

~~I would also like to have `begin_indent`/`end_indent` with ability to recursively do them for the same purpose. Currently I simulate them with such code:~~
<details>

```rust
fn indent_id() -> egui::Id {
    egui::Id::from("indent_id")
}

pub fn end_row_indent(ui: &mut egui::Ui) {
    ui.end_row();
    let count = ui.data_mut(|d| d.get_temp::<usize>(indent_id()).unwrap_or(0));
    if count != 0 {
        ui.monospace(std::iter::repeat_n("| ", count).collect::<String>());
    }
}

pub fn inc_indent(ui: &mut egui::Ui) {
    ui.data_mut(|d| {
        let count = d.get_temp::<usize>(indent_id()).unwrap_or(0);
        d.insert_temp(indent_id(), count + 1);
    });
}

pub fn dec_indent(ui: &mut egui::Ui) {
    ui.data_mut(|d| {
        let count = d.get_temp::<usize>(indent_id()).unwrap_or(0);
        d.insert_temp(indent_id(), count - 1);
    });
}
```

</details>

**UPD:** I just analyzed my code and came to conclusion, that for my very complex use-case I can survive if I have only such methods for `horizontal` and I can survive with such linear-type-ish API: (I would need `end_into_parent` method) (and I don't need custom begin/end for `indent` or `vertical` callback functions)
```rust
#[must_use = "Hold on to the returned UiScope until you are done adding widgets"]
pub struct UiScope<'a> {
  parent_ui: Option<&'a mut Ui>,
  child_ui: Option<Ui>,
  finalize: Option<UiScopeFinalize>,
}

impl<'a> UiScope<'a> {
  fn new(parent_ui: &'a mut Ui, child_ui: Ui, finalize: UiScopeFinalize) -> Self;

  pub fn ui(&mut self) -> &mut Ui;

  /// Finalize the scope and return the same [`Response`] as the callback-based helper.
  pub fn end(self) -> Response;

  /// Finalize the scope and return the parent [`Ui`] together with the resulting [`Response`].
  ///
  /// This is useful when a helper function may decide to end the current scope and immediately
  /// start another sibling scope on the same parent.
  pub fn end_into_parent(self) -> (&'a mut Ui, Response);
}

impl std::ops::Deref for UiScope<'_> {
  type Target = Ui;

  fn deref(&self) -> &Self::Target;
}

impl std::ops::DerefMut for UiScope<'_> {
  fn deref_mut(&mut self) -> &mut Self::Target;
}

impl Drop for UiScope<'_> {
  // Finish it, if user forgot
  fn drop(&mut self);
}

impl Ui {
  pub fn begin_horizontal(&mut self) -> UiScope<'_>;
}
```

@emilk if you fine with such interface for new `begin_horizontal` method and it deserves its own issue, please let me know! I would create it and maybe implement it.
